### ISSUE TITLE
Can not use winglibs inside pnpm packages

### ISSUE THREAD

**Issue by polamoros** [winglang/wing#5252]
### I tried this:

```
pnpm-workspace.yaml

packages:
  - packages/websockets
```

```
packages/websockets/main.w

bring websockets;
```

run:
````
cd packages/websockets/main.w
pnpm i @winglibs/websockets
wing it
```


### This happened:

```
Failed to compile.

error: Unable to load "cdktf": Module not found in "/Users/polamoros/Desktop/untitled folder/packages/websockets/node_modules/@winglibs/websockets/platform/tf-aws.w"
  --> node_modules/@winglibs/websockets/platform/tf-aws.w:4:1
  |
4 | bring "cdktf" as cdktf;
  | ^^^^^^^^^^^^^^^^^^^^^^^ Unable to load "cdktf": Module not found in "/Users/polamoros/Desktop/untitled folder/packages/websockets/node_modules/@winglibs/websockets/platform/tf-aws.w"


error: Unable to load "@cdktf/provider-aws": Module not found in "/Users/polamoros/Desktop/untitled folder/packages/websockets/node_modules/@winglibs/websockets/platform/tf-aws.w"
  --> node_modules/@winglibs/websockets/platform/tf-aws.w:5:1
  |
5 | bring "@cdktf/provider-aws" as tfaws;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Unable to load "@cdktf/provider-aws": Module not found in "/Users/polamoros/Desktop/untitled folder/packages/websockets/node_modules/@winglibs/websockets/platform/tf-aws.w"


error: Unable to load "aws-cdk-lib": Module not found in "/Users/polamoros/Desktop/untitled folder/packages/websockets/node_modules/@winglibs/websockets/platform/awscdk.w"
  --> node_modules/@winglibs/websockets/platform/awscdk.w:5:1
  |
5 | bring "aws-cdk-lib" as awscdk;
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Unable to load "aws-cdk-lib": Module not found in "/Users/polamoros/Desktop/untitled folder/packages/websockets/node_modules/@winglibs/websockets/platform/awscdk.w"
```

<img width="1075" alt="image" src="https://github.com/winglang/wing/assets/5547636/eab2267c-4cfa-42bb-a349-751e1a937382">


### I expected this:

to compile correctly as if I wasn't using it inside a pnpm package.

### Is there a workaround?

No other than installing all the dependencies of the winglib in the project root.

### Anything else?

_No response_

### Wing Version

0.51.17

### Node.js Version

_No response_

### Platform(s)

_No response_

### Community Notes

<!-- Please keep this note for the community -->
* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

**Comment by skyrpex** [comment]
Looks like something is off when resolving node dependencies.

**Comment by staycoolcall911** [comment]
Please ask Mark about it, he might have an idea.

**Comment by Boshen** [comment]
👋 

oxc_resolver is feature complete, you may give it a try.

> The main concern is getting something that works in WASI.

Work is on the way, you may track the progress here https://github.com/oxc-project/oxc_resolver/pull/31

**Comment by MarkMcCulloh** [comment]
Did some experimentation on this:

I utilized oxc_resolver for native builds (for unit tests) and exposed a `require.resolve` utility from the wasm host for WASI builds. This was able to resolve the problem for the compiler.

However, there is a fundamental incompatibility with PNPM after that: The JS we generate and execute gets _hoisted_ to a single target directory regardless of the source. This directory no longer has access to transitive dependencies at runtime (or bundle-time for inflights). This worked in the compiler because we can force resolution to happen from the source directory (e.g. from inside the `@winglibs/websockets` dir).

It may be possible to do something similar for the runtime with `createRequire` or code-genning require()s with overridden `paths` arguments. There are several problems with this that I did not continue with yet, as it is hostile to both bundling and packing. We could also exploring running js directly from node_modules and not recompiling them, but that's a pretty big lift. Then there is the question of ESM, which we'll want to switch to eventually.

#### TLDR

Will not continue working on this for now because this is specific to pnpm and there is a workaround: Either use `node-linker=hoisted` (boo) or directly add the transitive dependencies alongside the `@winglib` one

#### Side Note

About https://github.com/oxc-project/oxc_resolver/pull/31: oxc_resolver already technically worked in a WASI build (and has a great API btw), but symlink resolution must be disabled which obviously breaks usage for pnpm. The linked PR does not seem to address this issue, which makes sense since it's the responsibility the target runtime to provide `fs:canonicalize`. This makes me wonder if different WASI runtimes already fix this? Not sure, but worth exploring for other use-cases too.

**Comment by polamoros** [comment]
Hey,

I've been testing the workarounds you proposed and found issues with both..

With the `node-linker=hoisted` it only works for wing versions prior to `0.54.11`.
 It looks like the [`emit extern paths as relative`](https://github.com/winglang/wing/pull/5361) ruined the party..
```
Users/polamoros/Desktop/pnpm-wing-websockets/node_modules/@winglibs/websockets/inflight/websocket.aws.js must be a subdirectory of /Users/polamoros/Desktop/pnpm-wing-websockets/packages/websockets
```
I created a repo to test it [polamoros/pnpm-wing-websockets](https://github.com/polamoros/pnpm-wing-websockets), on the `use-hoisted-deps` branch.


Adding transitive dependencies is working with the demo project, but when I tried it in the wingcloud project, I ran into another error and I'm not sure where it's coming from..
```
error: Unknown symbol "KmsKey"
   --> crypto/crypto.aws.w:10:30
   |
10 |     let key = new aws.kmsKey.KmsKey();
   |                              ^^^^^^ Unknown symbol "KmsKey"

error: Unknown symbol "IamUser"
   --> runtime/runtime.w:298:37
   |
298 |       let awsUser = new aws.iamUser.IamUser(name: "{this.node.addr}-user");
   |    
```
I pushed the changes on `test-transitive-dependencies-workaround` branch of the [wingcloud repository](https://github.com/winglang/wing.cloud).


@MarkMcCulloh is there any workaround for the `hoisted` workaround?

**Comment by polamoros** [comment]
I finally got it working adding the transitive dependencies.

The "Unknown symbol "KmsKey" error is related to import (bring) alias collisions.

I'll open an issue

**Comment by Chriscbr** [comment]
Related: https://github.com/winglang/wing/issues/6129

**Comment by Boshen** [comment]
> 👋
> 
> oxc_resolver is feature complete, you may give it a try.
> 
> > The main concern is getting something that works in WASI.
> 
> ~Work is on the way, you may track the progress here [oxc-project/oxc-resolver#31](https://github.com/oxc-project/oxc-resolver/pull/31)~
> 
> EDIT: wasi support is done - https://www.npmjs.com/package/@oxc-resolver/binding-wasm32-wasi/

Wasi support is done - https://www.npmjs.com/package/@oxc-resolver/binding-wasm32-wasi/
